File: SkytilsMod-main/src/main/java/skytils/skytilsmod/features/impl/misc/damagesplash/graphics/colors/CustomColor.java
package skytils.skytilsmod.features.impl.misc.damagesplash.graphics.colors;

import net.minecraft.client.renderer.GlStateManager;
import org.apache.commons.codec.digest.DigestUtils;
import skytils.skytilsmod.utils.MathUtil;

import java.util.Arrays;

/** CustomColor
 * will represent color or complex colors
 * in a more efficient way than awt's Color or minecraft's color ints.
 *
 * Taken from Wynntils under GNU Affero General Public License v3.0
 * https://github.com/Wynntils/Wynntils/blob/development/LICENSE
 * @author Wynntils
 */
public class CustomColor {
    public float
            r,  // The RED   value of the color(0.0f -> 1.0f)
            g,  // The GREEN value of the color(0.0f -> 1.0f)
            b,  // The BLUE  value of the color(0.0f -> 1.0f)
            a;  // The ALPHA value of the color(0.0f -> 1.0f)

    public CustomColor(float r, float g, float b) { this(r, g, b, 1.0f); }

    public CustomColor(float r, float g, float b, float a) {
        this.r = r;
        this.g = g;
        this.b = b;
        this.a = a;
    }

    public CustomColor(int r, int g, int b) {
        this.r = r / 255f;
        this.g = g / 255f;
        this.b = b / 255f;
        this.a = 1.0f;
    }

    public CustomColor() {}

    public CustomColor(CustomColor c) { this(c.r, c.g, c.b, c.a); }

    /** applyColor
     * Will set the color to OpenGL's active color
     */
    public void applyColor() {
        GlStateManager.color(r, g, b, a);
    }

    public CustomColor setA(float a) {
        this.a = a;
        return this;
    }

    public static CustomColor fromString(String string, float a) {
        if (string == null) return new CustomColor(0, 0, 0, a);
        String withoutHash = string.trim();
        if (withoutHash.startsWith("#")) withoutHash = withoutHash.substring(1).trim();
        if (withoutHash.equals("")) return new CustomColor(0, 0, 0, a);
        if (withoutHash.length() == 6) {
            try {
                return fromInt(Integer.parseInt(withoutHash, 16), a);
            } catch (Exception ignored) { }
        } else if (withoutHash.length() == 3) {
            // "rgb" -> "rrggbb"
            try {
                int rgb = Integer.parseInt(withoutHash, 16);
                int r = ((rgb >> 8) & 0xF) * 0x11;
                int g = ((rgb >> 4) & 0xF) * 0x11;
                int b = (rgb & 0xF) * 0x11;
                return fromBytes((byte) r, (byte) g, (byte) b, a);
            } catch (Exception ignored) { }
        } else if (withoutHash.length() == 2) {
            // "vv" -> "vvvvvv"
            try {
                byte v = (byte) Integer.parseInt(withoutHash, 16);
                return fromBytes(v, v, v, a);
            } catch (Exception ignored) { }
        }
        byte[] hash = DigestUtils.sha1(string);
        return fromBytes(hash[0], hash[1], hash[2], a);
    }

    public static CustomColor fromHSV(float h, float s, float v, float a) {
        a = MathUtil.clamp(a, 0, 1);
        if (v <= 0) return new CustomColor(0, 0, 0, a);
        if (v > 1) v = 1;
        if (s <= 0) return new CustomColor(v, v, v, a);
        if (s > 1) s = 1;

        float vh = ((h % 1 + 1) * 6) % 6;

        int vi = MathUtil.fastFloor(vh);
        float v1 = v * (1 - s);
        float v2 = v * (1 - s * (vh - vi));
        float v3 = v * (1 - s * (1 - (vh - vi)));

        switch (vi) {
            case 0: return new CustomColor(v, v3, v1, a);
            case 1: return new CustomColor(v2, v, v1, a);
            case 2: return new CustomColor(v1, v, v3, a);
            case 3: return new CustomColor(v1, v2, v, a);
            case 4: return new CustomColor(v3, v1, v, a);
            default: return new CustomColor(v, v1, v2, a);
        }
    }

    /**
     * @return float[4]{ h, s, v, a }
     */
    public float[] toHSV() {
        float hue, saturation, value;
        float cmax = Math.max(Math.max(r, g), b);
        float cmin = Math.min(Math.min(r, g), b);

        value = cmax;
        saturation = cmax == 0 ? 0 : (cmax - cmin) / cmax;
        if (saturation == 0) {
            hue = 0;
        } else {
            float redc = (cmax - r) / (cmax - cmin);
            float greenc = (cmax - g) / (cmax - cmin);
            float bluec = (cmax - b) / (cmax - cmin);
            if (r == cmax) {
                hue = bluec - greenc;
            } else if (g == cmax) {
                hue = 2.0f + redc - bluec;
            } else {
                hue = 4.0f + greenc - redc;
            }
            hue = hue / 6.0f;
            if (hue < 0) {
                hue = hue + 1.0f;
            }
        }

        return new float[]{ hue, saturation, value, a };
    }

    /**
     * `c.toInt() & 0xFFFFFF` to get `0xRRGGBB` (without alpha)
     *
     * @return 0xAARRGGBB
     */
    public int toInt() {
        int r = (int) (Math.min(this.r, 1f) * 255);
        int g = (int) (Math.min(this.g, 1f) * 255);
        int b = (int) (Math.min(this.b, 1f) * 255);
        int a = (int) (Math.min(this.a, 1f) * 255);
        return (a << 24) | (r << 16) | (g << 8) | b;
    }

    /**
     * Construct a CustomColor from an int 0xAARRGGBB
     *
     * @param argb 32 bits with the most significant 8 being the value of the alpha channel, followed by rgb values
     * @return A custom colour such that fromInt(x).toInt() == x
     */
    public static CustomColor fromInt(int argb) {
        return fromInt(argb & 0xFFFFFF, ((argb >>> 24) & 0xFF) / 255f);
    }

    /**
     * Construct a CustomColor from an int 0xRRGGBB and an alpha value
     *
     * @param rgb 24 bits (Most significant 8 are ignored)
     * @param a Alpha value of colour
     * @return A custom colour such that fromInt(rgb, a).toInt() & 0xFFFFFF == rgb && fromInt(rgb, a).a == a
     */
    public static CustomColor fromInt(int rgb, float a) {
        return fromBytes((byte) (rgb >>> 16), (byte) (rgb >>> 8), (byte) rgb, a);
    }

    public static CustomColor fromBytes(byte r, byte g, byte b) {
        return fromBytes(r, g, b, 0);
    }

    public static CustomColor fromBytes(byte r, byte g, byte b, byte a) {
        return fromBytes(r, g, b, Byte.toUnsignedInt(a) / 255f);
    }

    public static CustomColor fromBytes(byte r, byte g, byte b, float a) {
        return new CustomColor(Byte.toUnsignedInt(r) / 255f, Byte.toUnsignedInt(g) / 255f, Byte.toUnsignedInt(b) / 255f, a);
    }

    /** HeyZeer0: this is = rgba(1,1,1,1) **/
    @Override
    public String toString() {
        return "rgba(" + r + "," + g + "," + b + "," + a +")";
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;

        if (other instanceof CustomColor) {
            CustomColor c = (CustomColor) other;
            return r == c.r && g == c.g && b == c.b && a == c.a;
        }
        return false;
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(new float[]{ r, g, b, a });
    }

    /* package-private */ static class SetBase extends CustomColor {
        SetBase(int rgb) {
            super((rgb >> 16) / 255.f, ((rgb >> 8) & 0xFF) / 255.f, (rgb & 0xFF) / 255.f, 1);
        }

        SetBase(float r, float g, float b, float a) {
            super(r, g, b, a);
        }

        // Prevent setA on global references. Create a copy with `new CustomColor(c)` first.
        @Override public CustomColor setA(float a) {
            new UnsupportedOperationException("Cannot set alpha of common color").printStackTrace();
            return this;
        }
    }

}

File: src/main/kotlin/skytils/skytilsmod/utils/graphics/colors/CustomColor.kt
/*
 * Skytils - Hypixel Skyblock Quality of Life Mod
 * Copyright (C) 2021 Skytils
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package skytils.skytilsmod.utils.graphics.colors

import net.minecraft.client.renderer.GlStateManager
import org.apache.commons.codec.digest.DigestUtils
import skytils.skytilsmod.utils.MathUtil

/** CustomColor
 * will represent color or complex colors
 * in a more efficient way than awt's Color or minecraft's color ints.
 *
 * Taken from Wynntils under GNU Affero General Public License v3.0
 * https://github.com/Wynntils/Wynntils/blob/development/LICENSE
 * @author Wynntils
 */
open class CustomColor {
    var r = 0f

    // The RED   value of the color(0.0f -> 1.0f)
    var g = 0f

    // The GREEN value of the color(0.0f -> 1.0f)
    var b = 0f

    // The BLUE  value of the color(0.0f -> 1.0f)
    var a // The ALPHA value of the color(0.0f -> 1.0f)
        = 0f

    @JvmOverloads
    constructor(r: Float, g: Float, b: Float, a: Float = 1.0f) {
        this.r = r
        this.g = g
        this.b = b
        this.a = a
    }

    constructor(r: Int, g: Int, b: Int) {
        this.r = r / 255f
        this.g = g / 255f
        this.b = b / 255f
        a = 1.0f
    }

    constructor()
    constructor(c: CustomColor) : this(c.r, c.g, c.b, c.a)

    /** applyColor
     * Will set the color to OpenGL's active color
     */
    open fun applyColor() {
        GlStateManager.color(r, g, b, a)
    }

    open fun setA(a: Float): CustomColor {
        this.a = a
        return this
    }

    /**
     * @return float[4]{ h, s, v, a }
     */
    open fun toHSV(): FloatArray {
        var hue: Float
        val saturation: Float
        val value: Float
        val cmax = r.coerceAtLeast(g).coerceAtLeast(b)
        val cmin = r.coerceAtMost(g).coerceAtMost(b)
        value = cmax
        saturation = if (cmax == 0f) 0F else (cmax - cmin) / cmax
        if (saturation == 0f) {
            hue = 0f
        } else {
            val redc = (cmax - r) / (cmax - cmin)
            val greenc = (cmax - g) / (cmax - cmin)
            val bluec = (cmax - b) / (cmax - cmin)
            hue = when {
                r == cmax -> {
                    bluec - greenc
                }
                g == cmax -> {
                    2.0f + redc - bluec
                }
                else -> {
                    4.0f + greenc - redc
                }
            }
            hue /= 6.0f
            if (hue < 0) {
                hue += 1.0f
            }
        }
        return floatArrayOf(hue, saturation, value, a)
    }

    /**
     * `c.toInt() & 0xFFFFFF` to get `0xRRGGBB` (without alpha)
     *
     * @return 0xAARRGGBB
     */
    open fun toInt(): Int {
        val r = (r.coerceAtMost(1f) * 255).toInt()
        val g = (g.coerceAtMost(1f) * 255).toInt()
        val b = (b.coerceAtMost(1f) * 255).toInt()
        val a = (a.coerceAtMost(1f) * 255).toInt()
        return a shl 24 or (r shl 16) or (g shl 8) or b
    }

    fun toInt(alpha: Int): Int {
        val r = (r.coerceAtMost(1f) * 255).toInt()
        val g = (g.coerceAtMost(1f) * 255).toInt()
        val b = (b.coerceAtMost(1f) * 255).toInt()
        val a = alpha.coerceAtMost(255)
        return a shl 24 or (r shl 16) or (g shl 8) or b
    }

    /** this is = rgba(255,255,255,255)  */
    override fun toString(): String {
        return "rgba(" + (r * 255).toInt() + "," + (g * 255).toInt() + "," + (b * 255).toInt() + "," + (a * 255).toInt() + ")"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other is CustomColor) {
            return r == other.r && g == other.g && b == other.b && a == other.a
        }
        return false
    }

    override fun hashCode(): Int {
        return floatArrayOf(r, g, b, a).contentHashCode()
    }

    /* package-private */
    open class SetBase : CustomColor {
        constructor(rgb: Int) : super((rgb shr 16) / 255f, (rgb shr 8 and 0xFF) / 255f, (rgb and 0xFF) / 255f, 1f)
        constructor(r: Float, g: Float, b: Float, a: Float) : super(r, g, b, a)

        // Prevent setA on global references. Create a copy with `new CustomColor(c)` first.
        override fun setA(a: Float): CustomColor {
            UnsupportedOperationException("Cannot set alpha of common color").printStackTrace()
            return this
        }
    }

    companion object {
        fun fromString(string: String?, a: Float): CustomColor {
            if (string == null) return CustomColor(0F, 0F, 0F, a)
            var withoutHash = string.trim { it <= ' ' }
            if (withoutHash.startsWith("#")) withoutHash = withoutHash.substring(1).trim { it <= ' ' }
            if (withoutHash == "") return CustomColor(0F, 0F, 0F, a)
            when (withoutHash.length) {
                6 -> {
                    try {
                        return fromInt(withoutHash.toInt(16), a)
                    } catch (ignored: Exception) {
                    }
                }
                3 -> {
                    // "rgb" -> "rrggbb"
                    try {
                        val rgb = withoutHash.toInt(16)
                        val r = (rgb shr 8 and 0xF) * 0x11
                        val g = (rgb shr 4 and 0xF) * 0x11
                        val b = (rgb and 0xF) * 0x11
                        return fromBytes(r.toByte(), g.toByte(), b.toByte(), a)
                    } catch (ignored: Exception) {
                    }
                }
                2 -> {
                    // "vv" -> "vvvvvv"
                    try {
                        val v = withoutHash.toInt(16).toByte()
                        return fromBytes(v, v, v, a)
                    } catch (ignored: Exception) {
                    }
                }
            }
            val hash = DigestUtils.sha1(string)
            return fromBytes(hash[0], hash[1], hash[2], a)
        }

        fun fromHSV(h: Float, s: Float, v: Float, a: Float): CustomColor {
            var s = s
            var v = v
            var a = a
            a = a.coerceIn(0f, 1f)
            if (v <= 0) return CustomColor(0F, 0F, 0F, a)
            if (v > 1) v = 1f
            if (s <= 0) return CustomColor(v, v, v, a)
            if (s > 1) s = 1f
            val vh = (h % 1 + 1) * 6 % 6
            val vi = MathUtil.fastFloor(vh.toDouble())
            val v1 = v * (1 - s)
            val v2 = v * (1 - s * (vh - vi))
            val v3 = v * (1 - s * (1 - (vh - vi)))
            return when (vi) {
                0 -> CustomColor(v, v3, v1, a)
                1 -> CustomColor(v2, v, v1, a)
                2 -> CustomColor(v1, v, v3, a)
                3 -> CustomColor(v1, v2, v, a)
                4 -> CustomColor(v3, v1, v, a)
                else -> CustomColor(v, v1, v2, a)
            }
        }

        /**
         * Construct a CustomColor from an int 0xAARRGGBB
         *
         * @param argb 32 bits with the most significant 8 being the value of the alpha channel, followed by rgb values
         * @return A custom colour such that fromInt(x).toInt() == x
         */
        fun fromInt(argb: Int): CustomColor {
            return fromInt(argb and 0xFFFFFF, (argb ushr 24 and 0xFF) / 255f)
        }

        /**
         * Construct a CustomColor from an int 0xRRGGBB and an alpha value
         *
         * @param rgb 24 bits (Most significant 8 are ignored)
         * @param a Alpha value of colour
         * @return A custom colour such that fromInt(rgb, a).toInt() & 0xFFFFFF == rgb && fromInt(rgb, a).a == a
         */
        fun fromInt(rgb: Int, a: Float): CustomColor {
            return fromBytes((rgb ushr 16).toByte(), (rgb ushr 8).toByte(), rgb.toByte(), a)
        }

        fun fromBytes(r: Byte, g: Byte, b: Byte, a: Byte): CustomColor {
            return fromBytes(r, g, b, java.lang.Byte.toUnsignedInt(a) / 255f)
        }

        @JvmOverloads
        fun fromBytes(r: Byte, g: Byte, b: Byte, a: Float = 0f): CustomColor {
            return CustomColor(
                java.lang.Byte.toUnsignedInt(r) / 255f,
                java.lang.Byte.toUnsignedInt(g) / 255f,
                java.lang.Byte.toUnsignedInt(b) / 255f,
                a
            )
        }
    }
}